File: typescript/src/app/channels.controller.ts
import { Body, Controller, Delete, Get, HttpCode, Inject, Param, Post, Query, Req, Res } from '@nestjs/common';
import type { Response } from 'express';

import { conflict, forbidden, invalid, notFound, notFoundChannel } from '../apierr/apierr.js';
import type { Channel, ChannelMember, ChannelMessage } from '../domain/domain.js';
import { firstUrl, preview, Role } from '../domain/domain.js';
import { IdFactory } from '../idgen/idgen.js';
import { build, errUnknownBefore, parseLimit } from '../paging/paging.js';
import {
  CHANNEL_READ_GATE,
  CHANNEL_ROLE_GATE,
  LINK_PREVIEWER,
  MEMBERSHIP_CACHE,
  MEMBERSHIP_WRITER,
  MESSAGE_POSTED_PUBLISHER,
  UNREAD_COUNTERS,
  type ChannelReadGate,
  type ChannelRoleGate,
  type LinkPreviewer,
  type MembershipCache,
  type MembershipWriter,
  type MessagePostedPublisher,
  type UnreadCounters,
} from '../seams/seams.js';
import { Store } from '../store/store.js';
import { currentUser, type RequestWithUser } from './identity.middleware.js';

interface ChannelDto {
  id: string;
  name: string;
  private: boolean;
  memberCount?: number;
  createdAt: string;
}

interface MembershipDto {
  channelId: string;
  userId: string;
  role: string;
  joinedAt: string;
}

interface ChannelMessageDto {
  id: string;
  channelId: string;
  senderId: string;
  text: string;
  attachmentIds: string[];
  linkPreviewTitle: string | null;
  createdAt: string;
}

function toChannelMessageDto(m: ChannelMessage, attachmentIds: string[]): ChannelMessageDto {
  return {
    id: m.id,
    channelId: m.channelId,
    senderId: m.senderId,
    text: m.text,
    attachmentIds,
    linkPreviewTitle: m.linkPreviewTitle,
    createdAt: m.createdAt.toISOString(),
  };
}

@Controller()
export class ChannelsController {
  constructor(
    private readonly store: Store,
    private readonly ids: IdFactory,
    @Inject(CHANNEL_READ_GATE) private readonly readGate: ChannelReadGate,
    @Inject(CHANNEL_ROLE_GATE) private readonly roleGate: ChannelRoleGate,
    @Inject(MEMBERSHIP_WRITER) private readonly membershipWriter: MembershipWriter,
    @Inject(MESSAGE_POSTED_PUBLISHER) private readonly publisher: MessagePostedPublisher,
    @Inject(LINK_PREVIEWER) private readonly linkPreviewer: LinkPreviewer,
    @Inject(MEMBERSHIP_CACHE) private readonly cache: MembershipCache,
    @Inject(UNREAD_COUNTERS) private readonly unread: UnreadCounters,
  ) {}

  @Post('channels')
  @HttpCode(201)
  async createChannel(@Req() req: RequestWithUser, @Body() body: { name?: string; private?: boolean }): Promise<ChannelDto> {
    const caller = currentUser(req);
    const name = body.name ?? '';
    if ([...name].length < 1 || [...name].length > 100) {
      throw invalid('channel:name:invalid', 'name must be 1–100 chars.');
    }
    const ch: Channel = { id: this.ids.create(), name, private: body.private ?? false, createdAt: new Date() };
    const owner: ChannelMember = { channelId: ch.id, userId: caller.id, role: Role.Owner, joinedAt: new Date() };
    await this.store.insertChannelWithOwner(ch, owner);
    return { id: ch.id, name: ch.name, private: ch.private, createdAt: ch.createdAt.toISOString() };
  }

  @Get('channels')
  async listChannels(@Req() req: RequestWithUser, @Query('limit') limit?: string, @Query('before') before?: string): Promise<unknown> {
    const caller = currentUser(req);
    const parsed = parseLimit(limit);
    const cursor = before ?? '';
    if (cursor !== '' && !(await this.store.channelExists(cursor))) {
      errUnknownBefore();
    }
    const rows = await this.store.visibleChannels(caller.id, cursor, parsed + 1);
    const dtos: ChannelDto[] = [];
    for (const c of rows) {
      const count = await this.store.memberCount(c.id);
      dtos.push({ id: c.id, name: c.name, private: c.private, memberCount: count, createdAt: c.createdAt.toISOString() });
    }
    return build(dtos, parsed, (d) => d.id);
  }

  @Get('channels/:id')
  async getChannel(@Req() req: RequestWithUser, @Param('id') id: string): Promise<ChannelDto> {
    const caller = currentUser(req);
    const ch = await this.readGate.authorizeRead(id, caller.id, false);
    const count = await this.store.memberCount(ch.id);
    return { id: ch.id, name: ch.name, private: ch.private, memberCount: count, createdAt: ch.createdAt.toISOString() };
  }

  @Post('channels/:id/join')
  @HttpCode(201)
  async joinChannel(@Req() req: RequestWithUser, @Param('id') channelId: string): Promise<MembershipDto> {
    const caller = currentUser(req);
    const ch = await this.store.channelById(channelId);
    if (!ch) {
      throw notFoundChannel();
    }
    const member = await this.store.membership(channelId, caller.id);
    if (member) {
      throw conflict('channel:member:already', 'You are already a member.');
    }
    if (ch.private) {
      throw notFoundChannel();
    }
    const m: ChannelMember = { channelId, userId: caller.id, role: Role.Member, joinedAt: new Date() };
    await this.membershipWriter.add(m);
    return { channelId, userId: caller.id, role: 'member', joinedAt: m.joinedAt.toISOString() };
  }

  @Post('channels/:id/members')
  @HttpCode(201)
  async addMember(@Req() req: RequestWithUser, @Param('id') channelId: string, @Body() body: { userId?: string }): Promise<MembershipDto> {
    const caller = currentUser(req);
    await this.roleGate.authorizeRole(channelId, caller.id, Role.Admin);
    const target = await this.store.userById(body.userId ?? '');
    if (!target) {
      throw notFound('user:not_found', 'User not found.');
    }
    const existing = await this.store.membership(channelId, target.id);
    if (existing) {
      throw conflict('channel:member:already', 'That user is already a member.');
    }
    const m: ChannelMember = { channelId, userId: target.id, role: Role.Member, joinedAt: new Date() };
    await this.membershipWriter.add(m);
    return { channelId, userId: target.id, role: 'member', joinedAt: m.joinedAt.toISOString() };
  }

  @Post('channels/:id/members/:userId/promote')
  @HttpCode(200)
  async promoteMember(@Req() req: RequestWithUser, @Param('id') channelId: string, @Param('userId') targetId: string): Promise<MembershipDto> {
    const caller = currentUser(req);
    await this.roleGate.authorizeRole(channelId, caller.id, Role.Owner);
    const target = await this.store.membership(channelId, targetId);
    if (!target) {
      throw notFound('channel:member:not_found', 'Member not found.');
    }
    if (target.role >= Role.Admin) {
      throw conflict('channel:member:already', 'That member is already an admin or owner.');
    }
    await this.store.updateMemberRole(channelId, targetId, Role.Admin);
    return { channelId, userId: targetId, role: 'admin', joinedAt: target.joinedAt.toISOString() };
  }

  @Delete('channels/:id/members/:userId')
  async removeMember(
    @Req() req: RequestWithUser,
    @Res() res: Response,
    @Param('id') channelId: string,
    @Param('userId') targetId: string,
  ): Promise<void> {
    const caller = currentUser(req);
    if (targetId === caller.id) {
      await this.leaveChannel(res, channelId, caller.id);
      return;
    }
    await this.kickMember(res, channelId, caller.id, targetId);
  }

  private async leaveChannel(res: Response, channelId: string, callerId: string): Promise<void> {
    const ch = await this.store.channelById(channelId);
    if (!ch) {
      throw notFoundChannel();
    }
    const member = await this.store.membership(channelId, callerId);
    if (!member) {
      if (ch.private) {
        throw notFoundChannel();
      }
      throw notFound('channel:member:not_found', 'Member not found.');
    }
    if (member.role === Role.Owner) {
      throw conflict('channel:owner:cannot_leave', 'The owner cannot leave their own channel.');
    }
    await this.membershipWriter.remove(channelId, callerId);
    res.status(204).send();
  }

  private async kickMember(res: Response, channelId: string, callerId: string, targetId: string): Promise<void> {
    const callerMembership = await this.roleGate.authorizeRole(channelId, callerId, Role.Admin);
    const target = await this.store.membership(channelId, targetId);
    if (!target) {
      throw notFound('channel:member:not_found', 'Member not found.');
    }
    const kickingPrivileged =
      target.role === Role.Owner || (target.role === Role.Admin && callerMembership.role !== Role.Owner);
    if (kickingPrivileged) {
      throw forbidden('channel:role:forbidden', 'Your role does not permit removing this member.');
    }
    await this.membershipWriter.remove(channelId, targetId);
    res.status(204).send();
  }

  @Delete('channels/:id')
  async deleteChannel(@Req() req: RequestWithUser, @Res() res: Response, @Param('id') channelId: string): Promise<void> {
    const caller = currentUser(req);
    await this.roleGate.authorizeRole(channelId, caller.id, Role.Owner);
    await this.store.deleteChannel(channelId);
    await this.cache.invalidate(channelId);
    res.status(204).send();
  }

  @Post('channels/:id/messages')
  async postChannelMessage(
    @Req() req: RequestWithUser,
    @Res() res: Response,
    @Param('id') channelId: string,
    @Body() body: { text?: string; attachmentIds?: string[] },
  ): Promise<void> {
    const caller = currentUser(req);
    await this.roleGate.authorizeRole(channelId, caller.id, Role.Member);
    const text = body.text ?? '';
    const attachmentIds = body.attachmentIds ?? [];
    if (text.length < 1 || [...text].length > 4000) {
      throw invalid('message:text:invalid', 'text must be 1–4000 chars.');
    }
    if (attachmentIds.length > 10) {
      throw invalid('message:attachment:invalid', 'A message can reference at most 10 attachments.');
    }
    await this.validateAttachments(channelId, caller.id, attachmentIds);

    // Unfurl runs (bounded, graceful) BEFORE the insert so the title persists with
    // the row; a slow/failing upstream degrades to null, never a hang (G-HTTP).
    let title: string | null = null;
    const url = firstUrl(text);
    if (url !== '') {
      title = await this.linkPreviewer.preview(url);
    }

    const msg: ChannelMessage = {
      id: this.ids.create(),
      channelId,
      senderId: caller.id,
      text,
      linkPreviewTitle: title,
      createdAt: new Date(),
    };

    // Pinned write ordering (02-api.md §3, no outbox): open tx → insert message →
    // publish AWAITING broker confirmation → commit. A publish failure rolls back
    // and surfaces 503; the message is never half-posted (G-KAFKA producer).
    await this.store.prisma.$transaction(async (tx) => {
      await this.store.insertChannelMessage(tx, msg);
      await this.store.attachMessageToAttachments(tx, msg.id, attachmentIds);
      await this.publisher.publish({
        messageId: msg.id,
        channelId,
        senderId: caller.id,
        preview: preview(text),
        postedAt: msg.createdAt.toISOString(),
      });
    });

    res.status(201).json(toChannelMessageDto(msg, attachmentIds));
  }

  // validateAttachments enforces S-AT-04: every referenced attachment must be
  // uploaded by the caller to THIS channel. A bad reference is 422 before any write.
  private async validateAttachments(channelId: string, callerId: string, ids: string[]): Promise<void> {
    if (ids.length === 0) {
      return;
    }
    const owned = await this.store.attachmentsOwnedInChannel(channelId, callerId, ids);
    if (owned.length !== ids.length) {
      throw invalid(
        'message:attachment:invalid',
        'Attachments must be uploaded to this channel by you and not already referenced.',
      );
    }
  }

  @Get('channels/:id/messages')
  async getChannelMessages(
    @Req() req: RequestWithUser,
    @Param('id') channelId: string,
    @Query('limit') limit?: string,
    @Query('before') before?: string,
  ): Promise<unknown> {
    const caller = currentUser(req);
    await this.readGate.authorizeRead(channelId, caller.id, true);
    const parsed = parseLimit(limit);
    const cursor = before ?? '';
    if (cursor !== '' && !(await this.store.channelMessageExists(channelId, cursor))) {
      errUnknownBefore();
    }
    const rows = await this.store.channelMessages(channelId, cursor, parsed + 1);
    const dtos = rows.map((m) => toChannelMessageDto(m, []));
    return build(dtos, parsed, (d) => d.id);
  }

  @Post('channels/:id/read')
  async markChannelRead(@Req() req: RequestWithUser, @Res() res: Response, @Param('id') channelId: string): Promise<void> {
    const caller = currentUser(req);
    await this.readGate.authorizeRead(channelId, caller.id, true);
    await this.unread.reset(caller.id, channelId);
    res.status(204).send();
  }
}

File: java/src/main/java/dev/drim/relay/app/ChannelsController.java
package dev.drim.relay.app;

import com.fasterxml.jackson.annotation.JsonProperty;
import dev.drim.relay.app.ChannelDtos.ChannelDto;
import dev.drim.relay.app.ChannelDtos.ChannelMessageDto;
import dev.drim.relay.app.ChannelDtos.MembershipDto;
import dev.drim.relay.domain.Entities;
import dev.drim.relay.domain.Role;
import dev.drim.relay.id.IdFactory;
import dev.drim.relay.seams.ChannelReadGate;
import dev.drim.relay.seams.ChannelRoleGate;
import dev.drim.relay.seams.MembershipCache;
import dev.drim.relay.seams.MembershipWriter;
import dev.drim.relay.seams.UnreadCounters;
import dev.drim.relay.store.ChannelMemberRepository;
import dev.drim.relay.store.ChannelMessageRepository;
import dev.drim.relay.store.ChannelRepository;
import dev.drim.relay.store.UserRepository;
import dev.drim.relay.store.entity.ChannelMemberRow;
import dev.drim.relay.store.entity.ChannelRow;
import dev.drim.relay.web.ApiException;
import dev.drim.relay.web.CurrentUser;
import dev.drim.relay.web.Paging;
import java.time.Instant;
import java.util.List;
import org.springframework.data.domain.Limit;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

/**
 * Channels: create/list/get + membership (join/add/promote/remove/leave/delete) + message read +
 * mark-read. The authorization decisions live behind {@link ChannelReadGate} (G-BOLA-READ) and
 * {@link ChannelRoleGate} (G-BOLA-ROLE); membership writes go through {@link MembershipWriter}
 * (G-CACHE). The handlers translate seam results into the pinned status codes.
 */
@RestController
public class ChannelsController {
  private final CurrentUser currentUser;
  private final ChannelRepository channels;
  private final ChannelMemberRepository members;
  private final ChannelMessageRepository messages;
  private final UserRepository users;
  private final ChannelReadGate readGate;
  private final ChannelRoleGate roleGate;
  private final MembershipWriter membershipWriter;
  private final MembershipCache cache;
  private final UnreadCounters unread;
  private final IdFactory ids;

  public ChannelsController(
      CurrentUser currentUser,
      ChannelRepository channels,
      ChannelMemberRepository members,
      ChannelMessageRepository messages,
      UserRepository users,
      ChannelReadGate readGate,
      ChannelRoleGate roleGate,
      MembershipWriter membershipWriter,
      MembershipCache cache,
      UnreadCounters unread,
      IdFactory ids) {
    this.currentUser = currentUser;
    this.channels = channels;
    this.members = members;
    this.messages = messages;
    this.users = users;
    this.readGate = readGate;
    this.roleGate = roleGate;
    this.membershipWriter = membershipWriter;
    this.cache = cache;
    this.unread = unread;
    this.ids = ids;
  }

  public record CreateChannelBody(String name, @JsonProperty("private") boolean isPrivate) {}

  public record AddMemberBody(String userId) {}

  @PostMapping("/channels")
  public ResponseEntity<ChannelDto> create(@RequestBody(required = false) CreateChannelBody body) {
    String name = body == null ? null : body.name();
    boolean isPrivate = body != null && body.isPrivate();
    if (name == null || name.isEmpty() || name.codePointCount(0, name.length()) > 100) {
      throw ApiException.invalid("channel:name:invalid", "name must be 1–100 chars.");
    }

    ChannelRow channel = new ChannelRow(ids.create(), name, isPrivate, Instant.now());
    channels.save(channel);
    members.save(
        new ChannelMemberRow(channel.getId(), currentUser.id(), Role.OWNER, Instant.now()));
    return ResponseEntity.status(HttpStatus.CREATED).body(ChannelDto.created(channel.toDomain()));
  }

  @GetMapping("/channels")
  public Paging.Page<ChannelDto> list(
      @RequestParam(required = false) String limit, @RequestParam(required = false) String before) {
    int parsedLimit = Paging.parseLimit(limit);
    if (before != null && !before.isEmpty() && !channels.existsById(before)) {
      throw Paging.unknownBefore();
    }
    List<ChannelDto> rows =
        channels
            .pageVisible(currentUser.id(), emptyToNull(before), fetchLimit(parsedLimit))
            .stream()
            .map(r -> ChannelDto.withCount(r.toDomain(), members.countByChannelId(r.getId())))
            .toList();
    return Paging.build(rows, parsedLimit, ChannelDto::id);
  }

  @GetMapping("/channels/{id}")
  public ChannelDto get(@PathVariable String id) {
    Entities.Channel channel = readGate.authorizeRead(id, currentUser.id(), false);
    return ChannelDto.withCount(channel, members.countByChannelId(id));
  }

  @PostMapping("/channels/{id}/join")
  public ResponseEntity<MembershipDto> join(@PathVariable String id) {
    Entities.Channel channel =
        channels.findById(id).map(r -> r.toDomain()).orElseThrow(NotFounds::channel);
    if (members.findByChannelIdAndUserId(id, currentUser.id()).isPresent()) {
      throw ApiException.conflict("channel:member:already", "You are already a member.");
    }
    if (channel.isPrivate()) {
      throw NotFounds.channel();
    }
    Instant joinedAt = Instant.now();
    membershipWriter.add(new Entities.ChannelMember(id, currentUser.id(), Role.MEMBER, joinedAt));
    return ResponseEntity.status(HttpStatus.CREATED)
        .body(new MembershipDto(id, currentUser.id(), Role.MEMBER.wire(), joinedAt));
  }

  @PostMapping("/channels/{id}/members")
  public ResponseEntity<MembershipDto> addMember(
      @PathVariable String id, @RequestBody(required = false) AddMemberBody body) {
    roleGate.authorizeRole(id, currentUser.id(), Role.ADMIN);
    String targetId = body == null ? null : body.userId();
    if (targetId == null || users.findById(targetId).isEmpty()) {
      throw ApiException.notFound("user:not_found", "User not found.");
    }
    if (members.findByChannelIdAndUserId(id, targetId).isPresent()) {
      throw ApiException.conflict("channel:member:already", "That user is already a member.");
    }
    Instant joinedAt = Instant.now();
    membershipWriter.add(new Entities.ChannelMember(id, targetId, Role.MEMBER, joinedAt));
    return ResponseEntity.status(HttpStatus.CREATED)
        .body(new MembershipDto(id, targetId, Role.MEMBER.wire(), joinedAt));
  }

  @PostMapping("/channels/{id}/members/{userId}/promote")
  public MembershipDto promote(@PathVariable String id, @PathVariable String userId) {
    roleGate.authorizeRole(id, currentUser.id(), Role.OWNER);
    Entities.ChannelMember target =
        members.findByChannelIdAndUserId(id, userId).map(r -> r.toDomain()).orElse(null);
    if (target == null) {
      throw ApiException.notFound("channel:member:not_found", "Member not found.");
    }
    if (target.role().atLeast(Role.ADMIN)) {
      throw ApiException.conflict(
          "channel:member:already", "That member is already an admin or owner.");
    }
    ChannelMemberRow row = members.findByChannelIdAndUserId(id, userId).orElseThrow();
    row.setRole(Role.ADMIN);
    members.save(row);
    return new MembershipDto(id, userId, Role.ADMIN.wire(), target.joinedAt());
  }

  @DeleteMapping("/channels/{id}/members/{userId}")
  public ResponseEntity<Void> removeMember(@PathVariable String id, @PathVariable String userId) {
    String callerId = currentUser.id();
    if (userId.equals(callerId)) {
      return leave(id, callerId);
    }
    return kick(id, callerId, userId);
  }

  private ResponseEntity<Void> leave(String channelId, String callerId) {
    Entities.Channel channel =
        channels.findById(channelId).map(r -> r.toDomain()).orElseThrow(NotFounds::channel);
    Entities.ChannelMember member =
        members.findByChannelIdAndUserId(channelId, callerId).map(r -> r.toDomain()).orElse(null);
    if (member == null) {
      if (channel.isPrivate()) {
        throw NotFounds.channel();
      }
      throw ApiException.notFound("channel:member:not_found", "Member not found.");
    }
    if (member.role() == Role.OWNER) {
      throw ApiException.conflict(
          "channel:owner:cannot_leave", "The owner cannot leave their own channel.");
    }
    membershipWriter.remove(channelId, callerId);
    return ResponseEntity.noContent().build();
  }

  private ResponseEntity<Void> kick(String channelId, String callerId, String targetId) {
    Entities.ChannelMember caller = roleGate.authorizeRole(channelId, callerId, Role.ADMIN);
    Entities.ChannelMember target =
        members.findByChannelIdAndUserId(channelId, targetId).map(r -> r.toDomain()).orElse(null);
    if (target == null) {
      throw ApiException.notFound("channel:member:not_found", "Member not found.");
    }
    boolean kickingPrivileged =
        target.role() == Role.OWNER || (target.role() == Role.ADMIN && caller.role() != Role.OWNER);
    if (kickingPrivileged) {
      throw ApiException.forbidden(
          "channel:role:forbidden", "Your role does not permit removing this member.");
    }
    membershipWriter.remove(channelId, targetId);
    return ResponseEntity.noContent().build();
  }

  @DeleteMapping("/channels/{id}")
  public ResponseEntity<Void> delete(@PathVariable String id) {
    roleGate.authorizeRole(id, currentUser.id(), Role.OWNER);
    channels.deleteById(id);
    cache.invalidate(id);
    return ResponseEntity.noContent().build();
  }

  @GetMapping("/channels/{id}/messages")
  public Paging.Page<ChannelMessageDto> messages(
      @PathVariable String id,
      @RequestParam(required = false) String limit,
      @RequestParam(required = false) String before) {
    readGate.authorizeRead(id, currentUser.id(), true);
    int parsedLimit = Paging.parseLimit(limit);
    if (before != null && !before.isEmpty() && !messages.existsByChannelIdAndId(id, before)) {
      throw Paging.unknownBefore();
    }
    List<ChannelMessageDto> rows =
        messages.page(id, emptyToNull(before), fetchLimit(parsedLimit)).stream()
            .map(r -> ChannelMessageDto.of(r.toDomain(), null))
            .toList();
    return Paging.build(rows, parsedLimit, ChannelMessageDto::id);
  }

  @PostMapping("/channels/{id}/read")
  public ResponseEntity<Void> markRead(@PathVariable String id) {
    readGate.authorizeRead(id, currentUser.id(), true);
    unread.reset(currentUser.id(), id);
    return ResponseEntity.noContent().build();
  }

  private static String emptyToNull(String s) {
    return (s == null || s.isEmpty()) ? null : s;
  }

  private static Limit fetchLimit(int limit) {
    return Limit.of(limit + 1);
  }
}
